pandas/_libs/src/vendored/numpy/datetime/np_datetime.c

```diff
@@ -482,10 +482,20 @@ npy_datetime npy_datetimestruct_to_datetime(NPY_DATETIMEUNIT base,
 
   if (base == NPY_FR_ns) {
     int64_t nanoseconds;
-    PD_CHECK_OVERFLOW(
-        scaleMicrosecondsToNanoseconds(microseconds, &nanoseconds));
-    PD_CHECK_OVERFLOW(
-        checked_int64_add(nanoseconds, dts->ps / 1000, &nanoseconds));
+
+    // for near-minimum timestamps, scaling microseconds to nanoseconds
+    // overflows but adding nanoseconds puts the timestamp back in a valid range
+    const int64_t min_nanoseconds = NPY_MIN_INT64 + 1;
+    if (microseconds == min_nanoseconds / 1000 - 1) {
+      // calculate final nanoseconds from minimum without scaling microseconds
+      PD_CHECK_OVERFLOW(checked_int64_add(
+          min_nanoseconds, (dts->ps - _NS_MIN_DTS.ps) / 1000, &nanoseconds));
```

Comment:
Can `(dts->ps - _NS_MIN_DTS.ps)` overflow? A comment as to what this is doing would be helpful

Reply:
I don't think it can overflow. `ps` is an `npy_int32`, and from the structure of `npy_datetimestruct`, the min/max values it should assume are `[-999999, 999999]` (just short of one microsecond, the next higher unit in `npy_datetimestruct`). So the min/max values of the subtraction here should be `[-1999998, 1999998]`, both of which fit well within `npy_int32`.

I'll add a comment on the operation. It offsets the nanoseconds by the smallest possible value because we're doing the calculation "in reverse".